### PR TITLE
Realtime publishers

### PR DESCRIPTION
**Pull request by rtnpro**
This implements publishers from which we can subscribe to data (say, latest message logs in a channel) in realtime.

### CODE REVIEW

**Comment by rtnpro** [comment]
@pypingou Now, this is what I was talking about... **realtime publishers**

**Comment by rtnpro** [comment]
@sayanchowdhury see, if you can run `ircb/publishers/__init__.py`?

File: ircb/publishers/__init__.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        if data['id'] in self.index:
+            self.index[data['id']] = data
+
+    def handle_create(self, data):
+        """
+        Check if an insert operation in message_logs table affects our
+        results. If yes, append it to results.
+        """
+        if self.skip(data):
+            return
```
Comment (pypingou) [inline ircb/publishers/__init__.py]: We logged something above (in handle_update), should we log here as well?

File: ircb/publishers/__init__.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        if self.results and data['timestamp'] < self.index[
+                self.results[-1]]['timestamp']:
+            return
+        if not self.fetched:
+            return
+
+        logger.debug('skip created data', data)
```
Comment (pypingou) [inline ircb/publishers/__init__.py]: This debug line seems out of place

File: ircb/publishers/__init__.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        logger.debug('updated results: %s, %s', self.results, self.index)
+
+    def skip(self, data):
+        """
+        We'll skip updating our results if the insert/update event
+        is not relevant to us.
+        """
+        return data['user_id'] != self.user_id or \
+            data['hostname'] != self.hostname or \
+            data['roomname'] != self.roomname
```
Comment (pypingou) [inline ircb/publishers/__init__.py]: We may want to consider the ordering here, put the most stringent filter first.

File: ircb/stores/logs.py [hunk 4]
```diff
@@ -36,6 +32,23 @@ def create(cls, hostname, roomname, message, event, timestamp,
...
+
+        if order_by:
+            for item in order_by:
+                if item.startswith('-'):
+                    qs = qs.order_by(getattr(MessageLog, item[1:]).desc())
+                else:
+                    qs = qs.order_by(getattr(MessageLog, item))
+
+        qs = qs.limit(limit)
+        return sorted(qs.all(), key=lambda item: getattr(item, 'timestamp'))
```
Comment (pypingou) [inline ircb/stores/logs.py]: The sorted here seems odd considering the order_by above.

Plus it makes you iterate through all the results while you could ask the DB to do it for you :)

**Comment by pypingou** [comment]
```
+        self.results = deque(maxlen=self.limit)
+        self.index = {}
```

`self.results` is thus limited in size but `self.index` doesn't seem to be nor does it seem to be cleaned anywhere. Won't this end up being a memory leak or some sort?

File: ircb/publishers/__init__.py (hunk 1 above)
Comment (rtnpro) [inline ircb/publishers/__init__.py]: Yes. That I missed :+1:

File: ircb/publishers/__init__.py (hunk 2 above)
Comment (rtnpro) [inline ircb/publishers/__init__.py]: Yes, my bad :(

File: ircb/publishers/__init__.py (hunk 3 above)
Comment (rtnpro) [inline ircb/publishers/__init__.py]: Yes. Currently we used the same `skip` function for both `handle_create` and `handle_update` and added some more skip conditions inside `handle_create`. May be we can have a method specific `skip_*` methods, as the filter for `update` and `create` events are going to be different.

File: ircb/stores/logs.py (hunk 4 above)
Comment (rtnpro) [inline ircb/stores/logs.py]: The problem statement here is that we want a list of latest N logs sorted by date:

filter()-> order_by('-date') -> limit(n) -> order_by('date')

AFAIK, I do not know how achieve the same via SqlAlchemy queries. I found a solution for this using RAW SQL queries: http://stackoverflow.com/a/2597100

Comment (pypingou) [inline ircb/stores/logs.py]: Using subquery would work indeed :)

File: ircb/publishers/__init__.py (hunk 3 above)
Comment (pypingou) [inline ircb/publishers/__init__.py]: I don't mind the design of a single `skip()` function I was merely pointing out that to optimize this list of `OR` conditions we should order the checks from the one that will return `False` most often to the one that will return `False` the least often :)

File: ircb/publishers/__init__.py (hunk 1 above)
Comment (pypingou) [inline ircb/publishers/__init__.py]: In that case we probably want to add logging for the other if as well while at it :)

File: ircb/publishers/__init__.py (hunk 3 above)
Comment (rtnpro) [inline ircb/publishers/__init__.py]: Agreed :+1:

File: ircb/stores/logs.py (hunk 4 above)
Comment (rtnpro) [inline ircb/stores/logs.py]: Ok, cool. I will try that out :D

**Comment by rtnpro** [comment]
@pypingou Yes, I know about the issue of `self.index` not getting cleaned. It seems like I have to do it manually everytime I add data to a queue which is full.

**Comment by rtnpro** [comment]
@pypingou Thanks for the review. I'd also like to know about your feedback on this concept of implementing a realtime publisher, powered by a dispatcher based store. That's what I meant when I said I was inspired by `meteor`, `flux`, `fedmsg`, hehe :) It did not turn up to be that ugly. What I am trying to get at is come up with a design pattern to make realtime stores, and in the long run, we can make it a standalone library, if we succeed with it in `waartaa`/`ircb`.

File: ircb/publishers/__init__.py (hunk 3 above)
Comment (rtnpro) [inline ircb/publishers/__init__.py]: Fixed!

File: ircb/publishers/__init__.py (hunk 2 above)
Comment (rtnpro) [inline ircb/publishers/__init__.py]: Fixed!

File: ircb/publishers/__init__.py (hunk 1 above)
Comment (rtnpro) [inline ircb/publishers/__init__.py]: Fixed!

File: ircb/stores/logs.py (hunk 4 above)
Comment (rtnpro) [inline ircb/stores/logs.py]: @pypingou, will
`return qs.from_self().order_by('anon_1_message_logs_timestamp').all()`

a sane way to do it?

Comment (pypingou) [inline ircb/stores/logs.py]: I guess I would adjust `qs` to `qs = session.query(MessageLog.id)` then do something like:

`return session.query(MessageLog).filter(MessageLog.id.in_(qs.subquery()).order_by(MessageLog.timestamp).all()`

**Comment by rtnpro** [comment]
Thanks, @pypingou for the review. I have fixed all the issues you pointed out. Merging it.

**Pull request merged**